openwisp_controller/subnet_division/rule_types/vpn.py

```diff
@@ -14,3 +15,15 @@ class VpnSubnetDivisionRuleType(BaseSubnetDivisionRuleType):
 
     organization_id_path = 'config.device.organization_id'
     subnet_path = 'vpn.subnet'
+
+    @staticmethod
+    def post_provision_handler(instance, provisioned, **kwargs):
+        def _assign_vpnclient_ip():
+            if provisioned['ip_addresses']:
+                instance.ip = provisioned['ip_addresses'][0]
+                instance.full_clean()
+                instance.save()
+
+        if not provisioned:
+            return
+        transaction.on_commit(_assign_vpnclient_ip)
```

Comment:
why then `transaction.on_commit` is called here as well? Is this intentional?
It's redundant, it does not look intentional to me. We should avoid redundancy in this case.

Reply:
This is intentional. From [Django's documentation](https://docs.djangoproject.com/en/3.2/topics/db/transactions/#performing-actions-after-commit):

> The function you pass in will be called immediately after a hypothetical database write made where on_commit() is called would be successfully committed.

[The previous](https://github.com/openwisp/openwisp-controller/pull/419#discussion_r630691666) `on_commit` will be fired when `VpnClient` is saved and the current one will be executed after `Subnet` and `IpAddress` objects have been saved in the database. 

Please let me know if my understanding of `transaction.on_commit` is wrong, because it will hurt us in the future again.

Reply:
as discussed we agreed that if anything fails in these operations we want everything to be rolledback so 1 transaction is enough.